File: 문제_2792번.cpp
#include <iostream>
#include <algorithm>
using namespace std;

//문제 2792번
typedef long long ll;
ll color[300001], total, n;
int main(){
    cin.tie(NULL);
    ios_base::sync_with_stdio(false);
    ll m,i , left = 1, right = -1 , mid, result =1000000001;
    cin >> n >> m;
    for(i=0;i<m;i++){
        cin >> color[i];
        right = max(color[i], right);//최대로 나누어 줄 수 있는 값을 지정 left는 최소 1개
    }
    while(left<=right){
        mid = (left + right)/2;//값을 맞춰가며 계속 나누어 줘본다
        total = 0;
        for(int i=0;i<m;i++){
            total += color[i] / mid;//보석을 일정 인원수 마다 나누어줄 수 있는지
            if(color[i] % mid)//나머지가 있으면
                total++;
        }
        if(n>=total){//구하는 값보다 이하값이면 되지만 아직 근접한 것은 아님
            result = min(result, mid);
            right = mid - 1;
        }
        else
            left = mid + 1;
    }
    cout << result;
    return 0;
}

File: JavaScript/문제_2792번.js
let input = require("fs")
  .readFileSync("/dev/stdin")
  .toString()
  .trim()
  .split("\n");
input[0] = input[0].split(' ');
let n = +input[0][0];
let m = +input[0][1];
input.shift();
let arr = [...input.map(o=>+o)];
//이분 탐색인데 result는 몇개로 나누어 줄 것 인가
let result = +1111111111;
let right = -1;
let left = +1;
for (let i = 0; i < arr.length; i++){
  right = Math.max(right, arr[i]);
}
let bigger = +0;
const divide = (val) => {
  let sum = +0;
  let maxBig = +0;
  for (let i = 0; i < arr.length; i++){
    sum += parseInt(arr[i] / val);
    if (parseInt(arr[i] % val))
      sum++;
  }
  if (sum <= n) {
    return true;
  }
  else
    false;
}
while (left <= right) {
  let mid = parseInt((left + right) / 2);
  bigger = +0;
  if (divide(mid)) {
    result = Math.min(result, mid);
    right = mid - 1;
  }
  else 
    left = mid + 1;
}
console.log(result);